File: python/kompute/basic/basic.py
import kp
import numpy as np

import os


def compile_source(file_name):
    os.system(f"glslangValidator -V {file_name} -o basic/basic.comp.spv")
    return open("basic/basic.comp.spv", "rb").read()


def main():
    mgr = kp.Manager()

    tensor_size = 4
    tensor_shape = [tensor_size, tensor_size]
    tensor_in_1 = mgr.tensor(np.triu(np.ones(tensor_shape)))
    tensor_in_2 = mgr.tensor(np.triu(np.ones(tensor_shape)))
    tensor_out = mgr.tensor(np.zeros(tensor_shape))

    print(f'Input tensors:\n'
          f'{tensor_in_1.data().reshape(tensor_shape)}\n'
          f'{tensor_in_2.data().reshape(tensor_shape)}\n')

    params = [tensor_in_1, tensor_in_2, tensor_out]

    matmul_shader = compile_source('''basic/basic.comp.glsl''')

    algo = mgr.algorithm(
        params,  # params
        matmul_shader,  # spirv
        (*tensor_shape, 1),  # workgroup
        [float(tensor_size)],  # spec_consts
        [])  # push_consts

    (mgr.sequence()
     .record(kp.OpTensorSyncDevice(params))
     .record(kp.OpAlgoDispatch(algo))
     .record(kp.OpTensorSyncLocal(params))
     .eval())

    print(f'Output :\n{tensor_out.data().reshape(tensor_shape)}')


if __name__ == '__main__':
    main()

File: cpp/kompute/basic/basic.cpp
#include <fstream>
#include <iostream>
#include <memory>
#include <vector>

#include "kompute/Kompute.hpp"

static std::vector<uint32_t> compile_source(const std::string& file_name) {
    std::string cmd =
        "glslangValidator -V " + file_name + " -o basic/basic.comp.spv";

    if (system(cmd.c_str())) {
        throw std::runtime_error("Error running glslangValidator command");
    }

    std::ifstream fileStream("basic/basic.comp.spv", std::ios::binary);
    std::vector<char> buffer;
    buffer
        .insert(buffer.begin(), std::istreambuf_iterator<char>(fileStream), {});

    return {
        (uint32_t*)buffer.data(),
        (uint32_t*)(buffer.data() + buffer.size())
    };
}

int main() {
    kp::Manager mgr;

    auto tensorInA = mgr.tensor({2.0, 4.0, 6.0});
    auto tensorInB = mgr.tensor({0.0, 1.0, 2.0});
    auto tensorOut = mgr.tensor({0.0, 0.0, 0.0});

    std::string shader("basic/basic.comp.glsl");

    std::vector<std::shared_ptr<kp::Tensor>> params =
        {tensorInA, tensorInB, tensorOut};

    std::shared_ptr<kp::Algorithm> algo =
        mgr.algorithm(params, compile_source(shader));

    mgr.sequence()
        ->record<kp::OpTensorSyncDevice>(params)
        ->record<kp::OpAlgoDispatch>(algo)
        ->record<kp::OpTensorSyncLocal>(params)
        ->eval();

    std::cout << "Output: {  ";
    for (const float& elem : tensorOut->vector()) {
        std::cout << elem << "  ";
    }
    std::cout << "}" << std::endl;

    return 0;
}
